### ISSUE TITLE
Adding comments in middle of a line

### ISSUE THREAD

**Issue by travisdowns** [skeeto/nasm-mode#7]
I'm not sure if this is a nasm-mode issue, or just a quirk of emacs comment handling, but when using the nasm mode, I can no longer add a `;` character in the middle of a line.

For example, I often have something like:

```
xor eax, dword [ebx + ecx + 123]
```

and I want to comment out everything after and including dword, like:

```
xor eax, ebx ; dword [ebx + ecx + 123]
```

When I enter the `;` in that location there, the cursor instead jumps to the end of the line and adds the ; there. In some cases this is helpful, but in this case not. I have to delete everything I want to comment out, and then write the comment.

nasm-mode issue or emacs quirk?

**Comment by skeeto** [comment]
This is by design. The ";" key is bound to nasm-comment. The function's
documentation explains all the behavior. You could rebind ";" to
self-insert-command to get simple behavior back. Or on a case-by-case
basis, you could do a literal insertion with "C-q ;".

I did this because assembly tends of have lots of gutter comments, and
this behavior felt natural to me.

**Comment by travisdowns** [comment]
OK, thanks! I guess `C-q ;` is what I was looking for, since the nasm-comment behavior is often useful as you point out (but sometimes I need the literal insertion). I didn't even think to read the function doc, but I'm a returning emacs user and it's coming back slowly. 

BTW, what's a gutter comment? Google failed me here...

Feel free to close this issue, of course.

**Comment by skeeto** [comment]
Considering that this doesn't show up in a Google search (I just tried
it myself), perhaps I'm using the wrong term. What I mean is the
right-hand column of comments often found in assembly listings. That's
where it jumps when you hit ";" in certain circumstances.

**Comment by travisdowns** [comment]
Makes sense. BTW is there any place the mode is documented that I should be reading (outside of the blog post that introduced it), or should I rely mostly on in-emacs docs?

**Comment by lassik** [comment]
The main reference for this stuff is the [*Comment Commands* section in the Emacs manual](https://www.gnu.org/software/emacs/manual/html_node/emacs/Comment-Commands.html).

The Emacs term for gutter comments seems to be *right-margin comments*. I got that from the docstring of the `comment-column` variable which sets the position of the right margin. *nasm-mode* implicitly uses that setting by calling `comment-indent`.

The standard behavior is for `;` to be a normal self-inserting character and for `M-;` to have "magic" behavior. `M-;` is globally bound to the command `comment-dwim` (it already works in *nasm-mode* as well, mostly as intended). It does something very similar to the special-purpose `nasm-comment` command that is bound to `;`.

Here's a comparison of `M-;` to the current `nasm-comment` command:

> * If no comment gutter exists yet, mark the current position and jump to it.

`M-;` does this (except it doesn't mark the current position).

> * If already within the gutter, pop the top mark and return to the code.

This is the one thing that `M-;` doesn't do - if you press it while inside a comment, it jumps to the beginning of the comment, not back to the code. I couldn't find a standard command to jump back to the code.

> * If on a line with no code, just insert a comment character.

`M-;` does this. (On blank lines the comment is not indented to the right margin by default, but is fixed by the code below.)

> * If within the indentation, just insert a comment character. This is intended prevent interference when the intention is to comment out the line.

This is usually done by just typing `;`.

> With a prefix arg, kill the comment on the current line with `comment-kill'."

`M-;` does this.

I tried to fix the behavior of `M-;` for NASM and here's the code:

```
(defun nasm-comment-indent ()
  comment-column)

(defun nasm-insert-comment ()
  (let ((comment-insert-comment-function nil))
    (comment-indent)))
```

And this to set buffer-local variables after entering *nasm-mode*:

```
(setq comment-indent-function 'nasm-comment-indent
      comment-insert-comment-function 'nasm-insert-comment)
```

What do you think about adopting this convention?

I also faintly recall there being a standard setting for an "electric" semicolon that auto-indents comments instead of self-inserting, but I can't find it now. I think at least Emacs Lisp mode had it on by default in some version of Emacs and that surprised me.

While trying this out, I realized that `nasm-indent-line` re-indents a comment on an otherwise-blank line to statement level, not to the right margin. I'm not sure whether you consider this a bug.

Sorry about the long-winded post. I thought to just leave a quick remark but ended up spending way too much time on it...

**Comment by lassik** [comment]
I packaged up the above code in PR #11 in any case since it improves `M-;` without breaking anything.

**Comment by skeeto** [comment]
> While trying this out, I realized that nasm-indent-line re-indents a 
> comment on an otherwise-blank line to statement level, not to the 
> right margin. I'm not sure whether you consider this a bug.

Yeah, I think that's a bug.
